### PR TITLE
Elasticsearch Activation Store

### PR DESCRIPTION
**Pull request by dubee**
## Description

Implementation for an Elasticsearch with Artifact activation store.

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [x] API
- [x] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [x] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [x] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

File: common/scala/src/main/scala/whisk/core/database/ArtifactElasticSearchActivationStore.scala [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+  }
+
+  override def get(activationId: ActivationId, context: UserContext)(
+    implicit transid: TransactionId): Future[WhiskActivation] = {
+    val headers = extractRequiredHeaders(context.request.headers)
+
+    // Return result from ElasticSearch or from artifact store if required headers are not present
+    if (headers.length == elasticSearchConfig.requiredHeaders.length) {
+      val uuid = elasticSearchConfig.path.format(context.user.namespace.uuid.asString)
+      val id = activationId.asString.substring(activationId.asString.indexOf("/") + 1)
```
Comment (dubee) [inline common/scala/src/main/scala/whisk/core/database/ArtifactElasticSearchActivationStore.scala]: I think something is wrong with `activationId` here, hence the substring.

File: common/scala/src/main/scala/whisk/core/database/ArtifactElasticSearchActivationStore.scala [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    val subject = Map("subject" -> activation.subject.toJson)
+    val activationId = Map("activationId" -> activation.activationId.toJson)
+    val start = Map("start" -> activation.start.toEpochMilli.toJson)
+    val end = Map("end" -> activation.end.toEpochMilli.toJson)
+    val cause = Map("cause" -> activation.cause.toJson)
+    val result = Map("result" -> activation.response.result.get.compactPrint.toJson)
+    val statusCode = Map("statusCode" -> activation.response.statusCode.toJson)
+    val logs = Map("logs" -> activation.logs.toJson)
+    val version = Map("version" -> activation.version.toJson)
+    val annotations = activation.annotations.toJsObject.fields
```
Comment (dubee) [inline common/scala/src/main/scala/whisk/core/database/ArtifactElasticSearchActivationStore.scala]: Annotations fields need to be lifted to the root level of the JSON object. Maybe there is a cleaner way.

File: common/scala/src/main/scala/whisk/core/database/ElasticSearchActivationRestClient.scala [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+        elasticSearchConfig.schema.activationId,
+        elasticSearchConfig.schema.version,
+        elasticSearchConfig.schema.start,
+        elasticSearchConfig.schema.end,
+        "result",
+        "statusCode",
+        elasticSearchConfig.schema.duration,
+        elasticSearchConfig.schema.namespace,
+        "kind",
+        "cause",
```
Comment (dubee) [inline common/scala/src/main/scala/whisk/core/database/ElasticSearchActivationRestClient.scala]: All of the hardcoded values for the JSON formatter need to be moved to configuration settings.

File: tests/src/test/scala/system/basic/WskRestBasicTests.scala [hunk 4]
```diff
@@ -1036,7 +1036,7 @@ class WskRestBasicTests extends TestHelpers with WskTestHelpers with WskActorSys
       result.getFieldJsValue("start").toString should not be JsObject.empty.toString
       result.getFieldJsValue("end").toString shouldBe JsObject.empty.toString
       result.getFieldJsValue("duration").toString shouldBe JsObject.empty.toString
-      result.getFieldListJsObject("annotations").length shouldBe 0
+      result.getFieldListJsObject("annotations").length shouldBe 1
```
Comment (dubee) [inline tests/src/test/scala/system/basic/WskRestBasicTests.scala]: Need to check if this works with the ArtifactStore.

File: tests/src/test/scala/whisk/core/containerpool/logging/ElasticSearchActivationStoreTests.scala [hunk 5]
```diff
@@ -0,0 +1,716 @@
...
+import whisk.common.TransactionId
+import whisk.core.entity.size.SizeInt
+
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future, Promise}
+import scala.util.{Success, Try}
+
+/*
+TODO:
+Required headers
```
Comment (dubee) [inline tests/src/test/scala/whisk/core/containerpool/logging/ElasticSearchActivationStoreTests.scala]: Need to address this TODO.

**Comment by style95** [comment]
+1 on this.

**Comment by ningyougang** [comment]
* The activation is not only saved into `elasticsearch`, but also saved into `couchdb`

I originally thought that the activation would only be saved into elasticsearch , and controller/invoker can get the activation from elasticsearch directly also.
So activation has no need to be saved into couchdb
How do you think about my originally thought?

* I save, you submited a pr : https://github.com/apache/incubator-openwhisk/pull/3991

Regarding `pull/3991`, it seems saved `activation` to `file and couchdb` and if want to save `activation`, should install logstash and elasticsearch by developers themselves.
Anywany, this patch and `pull/3991` have similar functions, what's the difference beween the two patches?

**Comment by dubee** [comment]
@style95, these changes will write activations to file and will allow the activation API in OpenWhisk to fetch activations from Elasticsearch. Changes not included here are the need for setting up Elasticsearch and a forwarder that passes the activations stored in files to Elasticsearch.

**Comment by style95** [comment]
@dubee Thanks for the clear answer.
BTW, is there any reason that storing activation in ES is not included in this PR?
Is it because of the performance?

**Comment by markusthoemmes** [comment]
This has been sitting in WIP, got stale and is inactive since 2 months. Any update?

**Comment by dubee** [comment]
Closing this as it is no longer applicable for what I need.

**Pull request closed without merging**